app/com/arpnetworking/metrics/portal/scheduling/impl/PeriodicSchedule.java

```diff
@@ -32,26 +36,68 @@
 public final class PeriodicSchedule extends BaseSchedule {
 
     private final ChronoUnit _period;
+    private final ZoneId _zone;
     private final Duration _offset;
 
     private PeriodicSchedule(final Builder builder) {
         super(builder);
         _period = builder._period;
+        _zone = builder._zone;
         _offset = builder._offset;
     }
 
     @Override
-    protected Optional<ZonedDateTime> unboundedNextRun(final Optional<ZonedDateTime> lastRun) {
-        if (!lastRun.isPresent()) {
-            ZonedDateTime alignedStart = getRunAtAndAfter().truncatedTo(_period);
-            if (alignedStart.isBefore(getRunAtAndAfter())) {
-                alignedStart = alignedStart.plus(Duration.of(1, _period));
-            }
-            return Optional.of(alignedStart.plus(_offset));
+    protected Optional<Instant> unboundedNextRun(final Optional<Instant> lastRun) {
+        final Instant runAfter = lastRun.orElse(getRunAtAndAfter().minus(Duration.ofNanos(1)));
+        return Optional.of(nextAlignedBoundaryAfter(runAfter).plus(_offset));
+    }
+
+    /**
+     * Returns the first instant after the given boundary that is [period]-aligned in our time zone.
+     *
+     * @param boundary The exclusive lower bound.
+     * @return A boundary that is guaranteed to be [period]-aligned and after the given boundary.
+     */
+    private Instant nextAlignedBoundaryAfter(final Instant boundary) {
+        final ZonedDateTime zonedLastRun = ZonedDateTime.ofInstant(boundary, _zone);
+        final ZonedDateTime truncated = zonedLastRun.truncatedTo(_period);
+        ZonedDateTime result = truncated.plus(_period.getDuration()).truncatedTo(_period);
+
+        // I am not intimately familiar with timekeeping standards, but:
+        //   it's conceivable that, for time-wonkiness reasons, the truncation might have moves us backwards
+        //   farther than adding `_period` moves us forward, meaning that `result` might still be too early.
+        // Just in case, and not very principled-ly, let's try larger steps.
+        if (!result.toInstant().isAfter(boundary)) {
+            result = truncated.plus(_period.getDuration().multipliedBy(2)).truncatedTo(_period);
+        }
+        if (!result.toInstant().isAfter(boundary)) {
+            result = truncated.plus(_period.getDuration().multipliedBy(3)).truncatedTo(_period);
```

Comment:
So it looks like we add 1 period above, then may add 2 more, and then 3 more for a total of 6 periods. In what cases might these be necessary? Why is it adding 1, 3 or 6 periods (e.g. as opposed to 1, 2, or 3)?

The truncate zeros all fields less than _period in magnitude. Adding durations of period should not impact truncation at all.

Failing to understand what we're concerned about. Is this for avoiding the DST change?

Reply:
> (I can't actually produce that failure mode, but my knowledge of TZDB is far from precise enough for me to state with confidence that nothing like this could possibly happen.)

Aha! I don't know why I couldn't produce this before, but I just wrote some code that tries 3 different implementations of "add one day and truncate to days", and they all fail catastrophically on a 25-hour day:
```
>>> import java.time.*
>>> import java.time.temporal.ChronoUnit
>>> 
>>> fun <T> printEquality(x: T, y: T) {
...     println("$x ${if (x==y) "==" else "!="} $y")
... }
>>> 
>>> val add24hAndTruncate = {zt: ZonedDateTime -> (zt + Duration.ofHours(24)).truncatedTo(ChronoUnit.DAYS)}
>>> val add1dAndTruncate = {zt: ZonedDateTime -> (zt + Duration.ofDays(1)).truncatedTo(ChronoUnit.DAYS)}
>>> val addPAndTruncate = {zt: ZonedDateTime, p: ChronoUnit -> (zt + p.duration).truncatedTo(p)}
>>> 
>>> val zt = ZonedDateTime.of(LocalDateTime.of(2018, 11, 4, 0, 0, 0), ZoneId.of("America/Los_Angeles"))
>>> 
>>> printEquality(zt, add24hAndTruncate(zt))
2018-11-04T00:00-07:00[America/Los_Angeles] == 2018-11-04T00:00-07:00[America/Los_Angeles]
>>> 
>>> printEquality(zt, add1dAndTruncate(zt))
2018-11-04T00:00-07:00[America/Los_Angeles] == 2018-11-04T00:00-07:00[America/Los_Angeles]
>>> 
>>> printEquality(zt, addPAndTruncate(zt, ChronoUnit.DAYS))
2018-11-04T00:00-07:00[America/Los_Angeles] == 2018-11-04T00:00-07:00[America/Los_Angeles]
```